### ISSUE TITLE
False positive of lifetime_capture_by

### ISSUE THREAD

**Issue by hokein** [llvm/llvm-project#121391]
See https://godbolt.org/z/Yb61Pda3W

```cpp
#include <vector>
struct Foo {};

template <typename T>
struct Container {
  const T& tt() [[clang::lifetimebound]];
};
template<typename T>
struct StatusOr {
   T* get() [[clang::lifetimebound]];
};
StatusOr<Container<const Foo*>> getContainer();

void test() {
  std::vector<const Foo*> vv;
  vv.push_back(getContainer().get()->tt()); // bogus diagnostic
}
```

**Comment by hokein** [comment]
Clang will automatically infer the `lifetime_capture_by` attribute for `vector`'s `push_back(const T& value)`.

When a vector is instantiated with a pointer type (`T` being `const Foo*`), the inferred annotation becomes `push_back(const Foo*& value [[clang::lifetime_capture_by(this)]])`.

For reference parameters, the `lifetime_capture_by` attribute treats the lifetime as referring to the referenced object -- in this case, the **pointer** itself, not the pointee object. In the `push_back`, we copy the pointer's value, which does not establish a reference to the pointer. This behavior is safe and does not capture the pointer's lifetime.

I think the `lifetime_capture_by` annotation is incorrectly applied here. It should not be inferred for cases where `T` is a pointer type, as the intended semantics do not align with the annotation. cc @usx95 @Xazax-hun

**Comment by usx95** [comment]
If we do not apply this to pointer types, we would miss when the pointer being stored is lifetimebound to a temporary.
```cpp
#include <vector>
struct Foo {};

template<typename T>
struct StatusOr {
   T* get() [[clang::lifetimebound]];
};
StatusOr<Foo> maybeGetFoo();

void test() {
  std::vector<Foo*> vv;
  vv.push_back(maybeGetFoo().get());
}
```
I think it should be fine to have such a false negative as this feels like a rare enough pattern.

**Comment by Xazax-hun** [comment]
One question is how to deal with unannotated types where we have no idea if they behave as a pointer or as an owner.

**Comment by usx95** [comment]
> One question is how to deal with unannotated types where we have no idea if they behave as a pointer or as an owner.

We currently infer capture_by for only gsl::Pointer and pointer types

**Comment by hokein** [comment]
> #include <vector>
> struct Foo {};
> 
> template<typename T>
> struct StatusOr {
>    T* get() [[clang::lifetimebound]];
> };
> StatusOr<Foo> maybeGetFoo();
> 
> void test() {
>   std::vector<Foo*> vv;
>   vv.push_back(maybeGetFoo().get());
> }

I think this is a nested case, which is not supported by design:

- ` push_back(const T & v [[lifetime_capture_by(this)]])`, T is a `Foo*`
- `lifetime_capture_by` only expresses the *top-level* lifetime, which is the referenced object, `Foo*`, not the pointee object `Foo`. For this case, there is no lifetime issue on the `Foo*`, thus no diagnostic.
